### PR TITLE
fix(uglify/worker): better `{Error}` handling

### PR DESCRIPTION
**Pull request by evilebottnawi**
It is happens only when parallel enable.

Before:
```shell
ERROR in vendor-8ab3e69853.js from UglifyJs
undefined
```
After:
```shell
ERROR in vendor-8ab3e69853.js from UglifyJs
TypeError: sym.definition is not a function
    at may_modify (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:8931:31)
    at TreeTransformer.abort [as before] (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:8541:74)
    at AST_VarDef.eval [as transform] (eval at <anonymous> /path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:4570:35)
    at eval (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:4586:25)
    at doit (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:130:23)
    at MAP (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:156:52)
    at do_list (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:4585:16)
    at eval (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:4664:28)
    at AST_Const.eval [as transform] (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:4573:17)
    at collapse (eval at <anonymous> (/path/to/node_modules/uglifyjs-webpack-plugin/node_modules/uglify-es/tools/node.js:21:1), <anonymous>:8617:39)
```

I can't do full testing using jest, because i can't found how mock modules in worker. Seems right now it is impossible, but you can test locally

Ref: https://github.com/rvagg/node-worker-farm/issues/9#issuecomment-20097934

### CODE REVIEW

File: src/uglify/worker.js [hunk 1]
```diff
@@ -7,7 +8,15 @@ module.exports = (options, callback) => {
...
-    callback(null, minify(options));
+    const result = minify(options);
+
+    // Communication channel employed by process.send() converts everything to JSON first.
+    // So custom object types (like Errors) can't possibly work, since object types aren't represented in JSON.
+    if (result.error) {
+      result.error = serializeError(result.error);
+    }
+
+    callback(null, result);
```
Comment (michael-ciniawsky) [inline src/uglify/worker.js]:
```js
if (result.error) cb(result.error)

cb(null, result)
```

?

```js
const path = require('path')

const postcss = require('postcss')
const postcssrc = require('postcss-load-config')

const SyntaxError = require('./SyntaxError')
const WorkerError = require('../WorkerError')

const css = (task, ctx, cb) => {
  const { plugins } = ctx
  const { file, data, options } = task

  return postcss(plugins)
    .process(data, options)
    .then(({ css, map }) => {
      cb(null, { file, css, map })
    })
    .catch((err) => {
      err.name === 'CssSyntaxError'
        ? cb(new SyntaxError(err))
        : cb(new WorkerError(err))
    })
}

module.exports = css
```

This works so far for `css-webpack-plugin`

Comment (michael-ciniawsky) [reply src/uglify/worker.js]: Whats is specific to `options.parallel` here ?

Comment (evilebottnawi) [reply src/uglify/worker.js]: @michael-ciniawsky i think about this, but for consistently don't use `error` in callback. Do you think better is use `callback(err, result)`?

`options.parallel` need for `worker-farm` when you use more than 1 process

Comment (michael-ciniawsky) [reply src/uglify/worker.js]: > Do you think better is use callback(err, result)?

Yep :), but I still wonder why the `{Error}` is malformed nevertheless, maybe the problem lies elsewhere...

Comment (evilebottnawi) [reply src/uglify/worker.js]: @michael-ciniawsky problem when you `process.send()`, it is call `JSON.stringify` and send data to other process (main), but `JSON.stringify` eats error(s). Just run `JSON.stringify({ error: new Error('test') });`

Comment (michael-ciniawsky) [reply src/uglify/worker.js]: **index.js** (CSS Plugin)
```js
workers.run(stylesheets, { plugins }, (errors, assets) => {
  if (errors && errors.length > 0) {
    errors.forEach((err) => {
      compilation.errors.push(err.message)
    })
}
```

**console.log(errors)**
```
[ { CSS Plugin:

      (7:3) Syntax Error: Unclosed block

        5 |   background-image: url('/assets/file.jpg');
        6 |   /* background-image: url('./file.png'); */
      > 7 | } {
          |   ^
        8 |
        9 |

          at postcss.process.then.catch (/Users/Cini/Github/Webpack/plugins/css-webpack-plugin/src/lib/worker/workers/postcss.js:22:14)
          at <anonymous>
          at runMicrotasksCallback (internal/process/next_tick.js:115:5)
          at process._tickCallback (internal/process/next_tick.js:150:11) type: 'SyntaxError', '$error': '$error', name: 'CSS Plugin' },
        { CSS Plugin:

      (7:3) Syntax Error: Unclosed block

        5 |   background-image: url('/assets/file.jpg');
        6 |   /* background-image: url('../../file.png'); */
      > 7 | } {
          |   ^
        8 |
        9 |

          at postcss.process.then.catch (/Users/Cini/Github/Webpack/plugins/css-webpack-plugin/src/lib/worker/workers/postcss.js:22:14)
          at <anonymous>
          at runMicrotasksCallback (internal/process/next_tick.js:115:5)
          at process._tickCallback (internal/process/next_tick.js:150:11) type: 'SyntaxError', '$error': '$error', name: 'CSS Plugin' },
        { CSS Plugin:

      (5:3) Syntax Error: Unclosed block

        3 |   background-image: url('/assets/file.jpg');
        4 |   /* background-image: url('../../../file.png'); */
      > 5 | } {
          |   ^
        6 |
        7 |

          at postcss.process.then.catch (/Users/Cini/Github/Webpack/plugins/css-webpack-plugin/src/lib/worker/workers/postcss.js:22:14)
          at <anonymous>
          at runMicrotasksCallback (internal/process/next_tick.js:115:5)
          at process._tickCallback (internal/process/next_tick.js:150:11) type: 'SyntaxError', '$error': '$error', name: 'CSS Plugin' },
        { CSS Plugin:

      (6:3) Syntax Error: Unclosed block

        4 |   background-image: url('/assets/file.jpg');
        5 |   /* background-image: url('../../../file.png'); */
      > 6 | } {
          |   ^
        7 |
        8 |

          at postcss.process.then.catch (/Users/Cini/Github/Webpack/plugins/css-webpack-plugin/src/lib/worker/workers/postcss.js:22:14)
          at <anonymous>
          at runMicrotasksCallback (internal/process/next_tick.js:115:5)
          at process._tickCallback (internal/process/next_tick.js:150:11) type: 'SyntaxError', '$error': '$error', name: 'CSS Plugin' },
        { CSS Plugin:

      (6:3) Syntax Error: Unclosed block

        4 |   background-image: url('/assets/file.jpg');
        5 |   /* background-image: url('../../../../file.png'); */
      > 6 | } {
          |   ^
        7 |
        8 |

          at postcss.process.then.catch (/Users/Cini/Github/Webpack/plugins/css-webpack-plugin/src/lib/worker/workers/postcss.js:22:14)
          at <anonymous>
          at runMicrotasksCallback (internal/process/next_tick.js:115:5)
          at process._tickCallback (internal/process/next_tick.js:150:11) type: 'SyntaxError', '$error': '$error', name: 'CSS Plugin' } ]
```

workererror

Comment (evilebottnawi) [reply src/uglify/worker.js]: @michael-ciniawsky `uglify-es` return only one error

Comment (michael-ciniawsky) [reply src/uglify/worker.js]: `postcss` also returns one `{Error}` :). I think this is the part for `{Errors}` I changed based on the worker implementation I initially took from `uglifyjs-webpack-plugin`


```js
const errors = [];
const results = [];

const next = (idx, result) => {
  tasks.pop()

  if (result.err) {
    errors[idx] = result.err;
  } else {
    results[idx] = result;
  }

  if (!tasks.length) {
    cb(errors, results);
  }
};
```

**Comment by evilebottnawi** [comment]
Close in favor https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/244

**Pull request closed without merging**